File: pkg/logger/logger.go
package logger

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"time"
)

// Stage 定義
const (
	StageReceive  = "receive"
	StageClassify = "classify"
	StageRoute    = "route"
	StageExecute  = "execute"
	StageResponse = "response"
	StageLog      = "log"
)

// Source 定義
const (
	SourceClawRouter = "clawrouter"
	SourceSmartLLM   = "smartllm"
	SourceTineye     = "tineye"
)

// Record 統一記錄格式
type Record struct {
	RequestID    string `json:"request_id"`
	Timestamp    string `json:"ts"`
	Source       string `json:"source"`
	Stage        string `json:"stage"`
	
	// 選填欄位
	UserMessage  string `json:"user_message,omitempty"`
	Tier         *int   `json:"tier,omitempty"`
	TierReason   string `json:"tier_reason,omitempty"`
	Model        string `json:"model,omitempty"`
	ActualModel  string `json:"actual_model,omitempty"`
	KeyID        string `json:"key_id,omitempty"`
	InputTokens  int    `json:"input_tokens,omitempty"`
	OutputTokens int    `json:"output_tokens,omitempty"`
	LatencyMs    int    `json:"latency_ms,omitempty"`
	Status       string `json:"status,omitempty"`
	ErrorMessage string `json:"error_message,omitempty"`
	FallbackTo   string `json:"fallback_to,omitempty"`
	Target       string `json:"target,omitempty"`
}

// Logger 實例
type Logger struct {
	collectorURL string
	source       string
	client       *http.Client
}

// New 建立新的 logger
func New(collectorURL, source string) *Logger {
	return &Logger{
		collectorURL: collectorURL,
		source:       source,
		client: &http.Client{
			Timeout: 3 * time.Second,
		},
	}
}

// GenerateRequestID 生成唯一 request ID
func GenerateRequestID() string {
	now := time.Now()
	random := fmt.Sprintf("%06x", rand.Intn(0xffffff))
	return fmt.Sprintf("req-%s-%s",
		now.Format("20060102-150405"),
		random)
}

// Log 記錄一筆資料
func (l *Logger) Log(record *Record) error {
	// 自動填入 timestamp 和 source
	if record.Timestamp == "" {
		record.Timestamp = time.Now().Format(time.RFC3339Nano)
	}
	if record.Source == "" {
		record.Source = l.source
	}

	data, err := json.Marshal(record)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", l.collectorURL, bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := l.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("collector returned %d", resp.StatusCode)
	}

	return nil
}

// LogReceive 記錄收到請求
func (l *Logger) LogReceive(requestID, userMessage string) error {
	return l.Log(&Record{
		RequestID:   requestID,
		Stage:       StageReceive,
		UserMessage: userMessage,
	})
}

// LogClassify 記錄分類結果
func (l *Logger) LogClassify(requestID string, tier int, reason, model string) error {
	return l.Log(&Record{
		RequestID:  requestID,
		Stage:      StageClassify,
		Tier:       &tier,
		TierReason: reason,
		Model:      model,
	})
}

// LogRoute 記錄路由決策
func (l *Logger) LogRoute(requestID, target string) error {
	return l.Log(&Record{
		RequestID: requestID,
		Stage:     StageRoute,
		Target:    target,
	})
}

// LogExecute 記錄執行開始
func (l *Logger) LogExecute(requestID, model, actualModel, keyID string, inputTokens int) error {
	return l.Log(&Record{
		RequestID:   requestID,
		Stage:       StageExecute,
		Model:       model,
		ActualModel: actualModel,
		KeyID:       keyID,
		InputTokens: inputTokens,
	})
}

// LogResponse 記錄回應完成
func (l *Logger) LogResponse(requestID string, outputTokens, latencyMs int, status string) error {
	return l.Log(&Record{
		RequestID:    requestID,
		Stage:        StageResponse,
		OutputTokens: outputTokens,
		LatencyMs:    latencyMs,
		Status:       status,
	})
}

// LogError 記錄錯誤
func (l *Logger) LogError(requestID, errorMessage string) error {
	return l.Log(&Record{
		RequestID:    requestID,
		Stage:        StageResponse,
		Status:       "error",
		ErrorMessage: errorMessage,
	})
}

File: pkg/logger/logger.py
"""TineyeLogger - Unified logging for OpenClaw Task Router"""
import json
import random
import time
from datetime import datetime, timezone
from typing import Optional
import aiohttp

# Stage 定義
STAGE_RECEIVE = "receive"
STAGE_CLASSIFY = "classify"
STAGE_ROUTE = "route"
STAGE_EXECUTE = "execute"
STAGE_RESPONSE = "response"
STAGE_LOG = "log"

# Source 定義
SOURCE_CLAWROUTER = "clawrouter"
SOURCE_SMARTLLM = "smartllm"
SOURCE_TINEYE = "tineye"


def generate_request_id() -> str:
    """生成唯一 request ID"""
    now = datetime.now()
    random_hex = format(random.randint(0, 0xffffff), '06x')
    return f"req-{now:%Y%m%d-%H%M%S}-{random_hex}"


class TineyeLogger:
    """統一 logger"""
    
    def __init__(self, collector_url: str, source: str):
        self.collector_url = collector_url
        self.source = source
    
    async def log(self, record: dict) -> None:
        """記錄一筆資料"""
        # 自動填入 timestamp 和 source
        if "ts" not in record:
            record["ts"] = datetime.now(timezone.utc).isoformat()
        if "source" not in record:
            record["source"] = self.source
        
        try:
            async with aiohttp.ClientSession() as session:
                await session.post(
                    self.collector_url,
                    data=json.dumps(record).encode(),
                    headers={"Content-Type": "application/json"},
                    timeout=aiohttp.ClientTimeout(total=3)
                )
        except Exception:
            pass  # 不阻塞主流程
    
    async def log_receive(self, request_id: str, user_message: str) -> None:
        """記錄收到請求"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_RECEIVE,
            "user_message": user_message[:500]  # 限制長度
        })
    
    async def log_classify(self, request_id: str, tier: int, reason: str, model: str) -> None:
        """記錄分類結果"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_CLASSIFY,
            "tier": tier,
            "tier_reason": reason,
            "model": model
        })
    
    async def log_route(self, request_id: str, target: str) -> None:
        """記錄路由決策"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_ROUTE,
            "target": target
        })
    
    async def log_execute(
        self,
        request_id: str,
        model: str,
        actual_model: str,
        key_id: str,
        input_tokens: int
    ) -> None:
        """記錄執行開始"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_EXECUTE,
            "model": model,
            "actual_model": actual_model,
            "key_id": key_id,
            "input_tokens": input_tokens
        })
    
    async def log_response(
        self,
        request_id: str,
        output_tokens: int,
        latency_ms: int,
        status: str = "success"
    ) -> None:
        """記錄回應完成"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_RESPONSE,
            "output_tokens": output_tokens,
            "latency_ms": latency_ms,
            "status": status
        })
    
    async def log_error(self, request_id: str, error_message: str) -> None:
        """記錄錯誤"""
        await self.log({
            "request_id": request_id,
            "stage": STAGE_RESPONSE,
            "status": "error",
            "error_message": error_message
        })
